### ISSUE TITLE
Some queries on `@apollo/gateway` 0.22 return "unreachable" error

### ISSUE THREAD

**Issue by jtomaszewski** [apollographql/federation#439]
After upgrading `@apollo/gateway` from 0.21 to 0.22, some queries work, and some queries simply don't, by returning weird `unreachable` error:

```js
// plugin that does the following:
// willSendResponse({ response }) {
//   console.log(response.errors, response.errors[0].extensions.exception);

[
  [ApolloError: unreachable] {
    locations: undefined,
    path: undefined,
    extensions: { code: 'INTERNAL_SERVER_ERROR', exception: [Object] }
  }
]

// response
{
  stacktrace: [
    'RuntimeError: unreachable',
    '    at wasm-function[525]:0xa0e4a',
    '    at wasm-function[583]:0xa3827',
    '    at wasm-function[718]:0xa6643',
    '    at wasm-function[688]:0xa6015',
    '    at wasm-function[711]:0xa64f6',
    '    at wasm-function[640]:0xa50df',
    '    at wasm-function[273]:0x89bf8',
    '    at wasm-function[32]:0x46b4e',
    '    at wasm-function[367]:0x94ce4',
    '    at wasm-function[31]:0x4563d'
  ]
}
```

What I've debugged so far:

- if I pass `debug: true` to `new ApolloGateway({ ... })`, it works
- if I downgrade gateway to ~0.21, it works

Any ideas what's wrong? Or how can I find the root cause of that error?

**Comment by queerviolet** [comment]
Hi, thanks for your patience. Do you have a repro case for this?

**Comment by abernix** [comment]
@jtomaszewski (and @nkreshchenko, based on your 👍 )  We'd like to look into this more.  If you can provide us with a small runnable reproduction that we can easily clone and run, that would be much appreciated.

**Comment by spencerarogers** [comment]
I ran into the same stacktrace on 0.23 with the following misconfiguration:

```js
const gateway = new ApolloGateway({
  // NOTE: Here the we have a service with a valid URL but a missing "name" property.
  serviceList: [{"service":"should-be-name", "url":"http://localhost:4001/graphql"}]
});
```

Changing "service" to "name" resolved my issue.

I was able to load the schema and docs in the GUI but then it would error out with that stacktrace when trying to execute a query.

Here's a reproduction: https://github.com/spencerarogers/apollo-federation-439-repro

**Comment by abernix** [comment]
@spencerarogers Thank you very much for finding that! This looks like it could be a reliable failure case which we should definitely fix, but I think it's probably distinct from the error that was being reported in this issue originally which was happening on some requests but not others.  (Unless @jtomaszewski and @nkreshchenko have a single service which is missing the `name` in the `serviceList` and it's only affecting queries which touch the unnamed service?)

Concretely, @spencerarogers: Do you mind opening a new issue on this repo with the reproduction you provided?

**Comment by spencerarogers** [comment]
@abernix Of course! I opened #473 for the `serviceList` `name` problem.

I ran a quick test on the scenario you describe in which there is a bad service configuration in a list of otherwise good service configurations and it seems to be the case that the gateway server will continue to function properly as long as it doesn't touch the improperly configured service and so I believe is still a possible explanation for the behavior in this issue. I did another reproduction with multiple services in this branch (which I also mention in #473):

https://github.com/spencerarogers/apollo-federation-439-repro/tree/multiple-implementing-services

**Comment by jtomaszewski** [comment]
Hmm in our case we use `name:` everywhere for the definitions of `serviceList`. So it might be different from #473 .

Unfortunately it's difficult for me to make a repro. Mainly because this error showed up to us only after deploy - locally it works fine...

I'll keep you updated if we have some progress on debugging this issue. (So far we rolled back to `@apollo/gateway` 0.22.0 ;/ )

**Comment by abernix** [comment]
Is this still happening on the latest version of Apollo Gateway?

**Comment by abernix** [comment]
I'm going to close this because I believe it's resolved.  Happy to reopen if it's still occurring on the latest version!
